src/ints/mouse.cpp

```diff
@@ -18,1549 +18,932 @@
+    {
+        return HasEventDOSbtn() && !delay_dos_btn &&
+               !mouse_shared.dos_cb_running; // callback busy = no new
+                                             // event
+    }
+    inline bool HasReadyEventAny() const
+    {
+        return HasReadyEventPS2() || HasReadyEventDOSmov() ||
+               HasReadyEventDOSbtn();
+    }
+
+} queue;
+
+static void MouseQueueTick(uint32_t)
 {
-	float dx = xrel * mouse.pixelPerMickey_x;
-	float dy = yrel * mouse.pixelPerMickey_y;
-
-	if ((fabs(xrel) > 1.0) || (mouse.senv_x < 1.0))
-		dx *= mouse.senv_x;
-	if ((fabs(yrel) > 1.0) || (mouse.senv_y < 1.0))
-		dy *= mouse.senv_y;
-	if (useps2callback)
-		dy *= 2;
-
-	mouse.mickey_x += (dx * mouse.mickeysPerPixel_x);
-	mouse.mickey_y += (dy * mouse.mickeysPerPixel_y);
-	if (mouse.mickey_x >= 32768.0)
-		mouse.mickey_x -= 65536.0;
-	else if (mouse.mickey_x <= -32769.0)
-		mouse.mickey_x += 65536.0;
-	if (mouse.mickey_y >= 32768.0)
-		mouse.mickey_y -= 65536.0;
-	else if (mouse.mickey_y <= -32769.0)
-		mouse.mickey_y += 65536.0;
-	if (emulate) {
-		mouse.x += dx;
-		mouse.y += dy;
-	} else {
-		if (CurMode->type == M_TEXT) {
-			mouse.x = x * real_readw(BIOSMEM_SEG, BIOSMEM_NB_COLS) * 8;
-			mouse.y = y * (IS_EGAVGA_ARCH ? (real_readb(BIOSMEM_SEG, BIOSMEM_NB_ROWS) + 1) : 25) * 8;
-		} else if ((mouse.max_x < 2048) || (mouse.max_y < 2048) ||
-		           (mouse.max_x != mouse.max_y)) {
-			if ((mouse.max_x > 0) && (mouse.max_y > 0)) {
-				mouse.x = x * mouse.max_x;
-				mouse.y = y * mouse.max_y;
-			} else {
-				mouse.x += xrel;
-				mouse.y += yrel;
-			}
-		} else { // Games faking relative movement through absolute
-			 // coordinates. Quite surprising that this actually
-			 // works..
-			mouse.x += xrel;
-			mouse.y += yrel;
-		}
-	}
-
-	/* ignore constraints if using PS2 mouse callback in the bios */
-
-	if (!useps2callback) {
-		if (mouse.x > mouse.max_x)
-			mouse.x = mouse.max_x;
-		if (mouse.x < mouse.min_x)
-			mouse.x = mouse.min_x;
-		if (mouse.y > mouse.max_y)
-			mouse.y = mouse.max_y;
-		if (mouse.y < mouse.min_y)
-			mouse.y = mouse.min_y;
-	} else {
-		if (mouse.x >= 32768.0)
-			mouse.x -= 65536.0;
-		else if (mouse.x <= -32769.0)
-			mouse.x += 65536.0;
-		if (mouse.y >= 32768.0)
-			mouse.y -= 65536.0;
-		else if (mouse.y <= -32769.0)
-			mouse.y += 65536.0;
-	}
-	AddEvent(EventType::MouseHasMoved);
-	DrawCursor();
+    queue.Tick();
 }
 
-static uint8_t GetResetWheel8bit()
-{
-	if (!mouse.cute_mouse)
-		return 0;
-	const int8_t tmp = static_cast<int8_t>(
-	        std::clamp(mouse.wheel,
-	                   static_cast<int16_t>(INT8_MIN),
-	                   static_cast<int16_t>(INT8_MAX)));
-	mouse.wheel = 0; // clear the wheel counter after reading
-	return (tmp >= 0) ? tmp : 0x100 + tmp; // 0xff represents -1, 0xfe
-	                                       // represents -2, etc.
-}
+MouseQueue::MouseQueue() {}
 
-static uint16_t GetResetWheel16bit()
+void MouseQueue::AddEvent(MouseEvent &event)
```

Comment:
const

Reply:
`AddEvent` changes `event` flags in some places, fills in aggregation bit mask - const would meant the necessity of creating additional structure copy.